src/runtime_src/core/pcie/common/asd_parser.h

```diff
@@ -0,0 +1,294 @@
+    XAIE_CORE_STATUS_LOCK_STALL_S_BIT,
+    XAIE_CORE_STATUS_LOCK_STALL_W_BIT,
+    XAIE_CORE_STATUS_LOCK_STALL_N_BIT,
+    XAIE_CORE_STATUS_LOCK_STALL_E_BIT,
+    XAIE_CORE_STATUS_STREAM_STALL_SS0_BIT,
+    XAIE_CORE_STATUS_STREAM_STALL_MS0_BIT = 12U,
+    XAIE_CORE_STATUS_CASCADE_STALL_SCD_BIT = 14U,
+    XAIE_CORE_STATUS_CASCADE_STALL_MCD_BIT,
+    XAIE_CORE_STATUS_DEBUG_HALT_BIT,
+    XAIE_CORE_STATUS_ECC_ERROR_STALL_BIT,
+    XAIE_CORE_STATUS_ECC_SCRUBBING_STALL_BIT,
+    XAIE_CORE_STATUS_ERROR_HALT_BIT,
+    XAIE_CORE_STATUS_DONE_BIT,
+    XAIE_CORE_STATUS_PROCESSOR_BUS_STALL_BIT,
+    XAIE_CORE_STATUS_MAX_BIT
+  };
+  
+  enum class dma_s2mm_status : uint32_t
+  {
+    XAIE_DMA_STATUS_S2MM_STATUS = 0U,
+    XAIE_DMA_STATUS_S2MM_STALLED_LOCK_ACK = 2U,
+    XAIE_DMA_STATUS_S2MM_STALLED_LOCK_REL,
+    XAIE_DMA_STATUS_S2MM_STALLED_STREAM_STARVATION,
+    XAIE_DMA_STATUS_S2MM_STALLED_TCT_OR_COUNT_FIFO_FULL,
+    XAIE_DMA_STATUS_S2MM_ERROR_LOCK_ACCESS_TO_UNAVAIL = 8U, // Specific only to MEM Tile
+    XAIE_DMA_STATUS_S2MM_ERROR_DM_ACCESS_TO_UNAVAIL,       // Specific only to MEM Tile
+    XAIE_DMA_STATUS_S2MM_ERROR_BD_UNAVAIL = 10U,
+    XAIE_DMA_STATUS_S2MM_ERROR_BD_INVALID,
+    XAIE_DMA_STATUS_S2MM_ERROR_FOT_LENGTH,
+    XAIE_DMA_STATUS_S2MM_ERROR_FOT_BDS_PER_TASK,
+    XAIE_DMA_STATUS_S2MM_AXI_MM_DECODE_ERROR = 16U,
+    XAIE_DMA_STATUS_S2MM_AXI_MM_SLAVE_ERROR  = 17U,
+    XAIE_DMA_STATUS_S2MM_TASK_QUEUE_OVERFLOW = 18U,
+    XAIE_DMA_STATUS_S2MM_CHANNEL_RUNNING,
+    XAIE_DMA_STATUS_S2MM_TASK_QUEUE_SIZE,
+    XAIE_DMA_STATUS_S2MM_CURRENT_BD = 24U,
+    XAIE_DMA_STATUS_S2MM_MAX
+  };
+  
+  enum class dma_mm2s_status : uint32_t
+  {
+    XAIE_DMA_STATUS_MM2S_STATUS = 0U,
+    XAIE_DMA_STATUS_MM2S_STALLED_LOCK_ACK = 2U,
+    XAIE_DMA_STATUS_MM2S_STALLED_LOCK_REL,
+    XAIE_DMA_STATUS_MM2S_STALLED_STREAM_BACKPRESSURE,
+    XAIE_DMA_STATUS_MM2S_STALLED_TCT,
+    XAIE_DMA_STATUS_MM2S_ERROR_LOCK_ACCESS_TO_UNAVAIL = 8U, // Specific only to MEM Tile
+    XAIE_DMA_STATUS_MM2S_ERROR_DM_ACCESS_TO_UNAVAIL,        // Specific only to MEM Tile
+    XAIE_DMA_STATUS_MM2S_ERROR_BD_UNAVAIL,
+    XAIE_DMA_STATUS_MM2S_ERROR_BD_INVALID = 11U,
+    XAIE_DMA_STATUS_MM2S_AXI_MM_DECODE_ERROR = 16U,
+    XAIE_DMA_STATUS_MM2S_AXI_MM_SLAVE_ERROR  = 17U,
+    XAIE_DMA_STATUS_MM2S_TASK_QUEUE_OVERFLOW = 18U,
+    XAIE_DMA_STATUS_MM2S_CHANNEL_RUNNING,
+    XAIE_DMA_STATUS_MM2S_TASK_QUEUE_SIZE,
+    XAIE_DMA_STATUS_MM2S_CURRENT_BD = 24U,
+    XAIE_DMA_STATUS_MM2S_MAX
+  };
+
+  void
+  aie_status_version_check(uint16_t &major_ver, uint16_t &minor_ver);
+
+  void
+  aie_info_sanity_check(uint32_t start_col, uint32_t num_cols, aie_tiles_info &info);
+
+  template <typename tile_type>
+  std::vector<std::unique_ptr<aie_col_status>>
+  parse_data_from_buf(char *buf, aie_tiles_info &info)
+  {
+    std::vector<std::unique_ptr<aie_col_status>> aie_cols;
+
+    for (uint32_t i = 0; i < info.cols; i++) {
+      aie_cols.emplace_back(std::make_unique<aie_col_status>(info));
```

Comment:
Is there any reason the vector elements are heap allocated and stored in unique_ptr?  I mean, as opposed to just use `aie_col_status` directly?

Reply:
yes @stsoe, to avoid unnecessary copies of aie_col_status I am creating this way.